trace/tracestate/tracestate.go

```diff
@@ -0,0 +1,168 @@
+	valueFormat            = `[\x20-\x2b\x2d-\x3c\x3e-\x7e]{0,255}[\x21-\x2b\x2d-\x3c\x3e-\x7e]`
+)
+
+var keyValidationRegExp = regexp.MustCompile(`^(` + keyFormat + `)$`)
+var valueValidationRegExp = regexp.MustCompile(`^(` + valueFormat + `)$`)
+
+// Tracestate represents tracing-system specific context in a list of key-value pairs. Tracestate allows different
+// vendors propagate additional information and inter-operate with their legacy Id formats.
+type Tracestate struct {
+	entries []Entry
+}
+
+// Entry represents one key-value pair in a list of key-value pair of Tracestate.
+// Key is an opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+// and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+// forward slashes /.
+//
+// Value is an opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+// range 0x20 to 0x7E) except comma , and =.
+type Entry struct {
+	Key   string
+	Value string
+}
+
+// Get retrieves value for a given key from Tracestate ts.
+// If the key is not found then false is returned with the value "".
+// If the key is found then true is returned with its value.
+func (ts *Tracestate) Get(key string) (string, bool) {
+	if len(ts.entries) == 0 {
+		return "", false
+	}
+	for _, entry := range ts.entries {
+		if entry.Key == key {
+			return entry.Value, true
+		}
+	}
+	return "", false
+}
+
+// Entries returns a slice of Entry.
+func (ts *Tracestate) Entries() []Entry {
+	return ts.entries
+}
+
+func (ts *Tracestate) remove(key string) *Entry {
+	for index, entry := range ts.entries {
+		if entry.Key == key {
+			ts.entries = append(ts.entries[:index], ts.entries[index+1:]...)
+			return &entry
+		}
+	}
+	return nil
+}
+
+func (ts *Tracestate) add(entry Entry) error {
+	ts.remove(entry.Key)
+	if len(ts.entries) >= maxKeyValuePairs {
+		return fmt.Errorf("reached maximum key/value pairs limit of %d", maxKeyValuePairs)
+	}
+	ts.entries = append([]Entry{entry}, ts.entries...)
+	return nil
+}
+
+func isValid(entry Entry) bool {
+	return keyValidationRegExp.MatchString(entry.Key) &&
+		valueValidationRegExp.MatchString(entry.Value)
+}
+
+func containsDuplicateKey(entries []Entry) (string, bool) {
+	keyMap := make(map[string]int)
+	for _, entry := range entries {
+		if _, ok := keyMap[entry.Key]; ok {
+			return entry.Key, true
+		}
+		keyMap[entry.Key] = 1
+	}
+	return "", false
+}
+
+func areEntriesValid(entries []Entry) (*Entry, bool) {
+	for _, entry := range entries {
+		if !isValid(entry) {
+			return &entry, false
+		}
+	}
+	return nil, true
+}
+
+// NewFromEntries creates a Tracestate object from an array of key-value pair.
+// nil is returned with an error if
+//  1. the len of the entries > maxKeyValuePairs
+//  2. the entries contain duplicate keys
+//  3. one or more entries are invalid.
+func NewFromEntries(entries []Entry) (*Tracestate, error) {
+	if len(entries) == 0 || len(entries) > maxKeyValuePairs {
```

Comment:
Why is 0 an invalid number of entries?

Reply:
I recommend that we consider `nil` to represent a valid, empty Tracestate as I mentioned in my previous comment. Then we can simply return `nil, nil`, if `len(entries) == 0`.

Reply:
nil is already a valid Tracestate. I'll remove the len == 0 check.